src/workerd/io/actor-cache.c++

```diff
@@ -2313,17 +2321,37 @@ kj::Promise<void> ActorCache::flushImpl(uint retryCount) {
     batch.wordCount += words;
   };
 
-  kj::Vector<CountedDeleteFlush> countedDeleteFlushes;
-
-  auto countEntry = [&](Entry& entry) {
-    // Counts up the number of operations and RPC message sizes we'll need to cover this entry.
-
-    entry.setFlushing();
+  kj::Vector<CountedDeleteFlush> countedDeleteFlushes(countedDeletes.size());
+  for (auto countedDelete : countedDeletes) {
+    if (countedDelete->isFinished) {
+      // This countedDelete has already be executed, but we haven't delivered the final count to
+      // the waiter yet. We'll skip it here since the destructor of CountedDeleteWaiter should
+      // eventually remove this entry from `countedDeletes`.
+      continue;
+    }
 
-    auto keySizeInWords = bytesToWordsRoundUp(entry.key.size());
+    // We might have successfully deleted these entries, but had the broader transaction fail.
+    // In that case, we might have entries that have since been overwritten, and which no longer
+    // need to be scheduled for deletion.
+    kj::Vector<kj::Own<Entry>> entriesToDelete(countedDelete->entries.size());
+    for (auto& entry : countedDelete->entries) {
+      if (entry->overwritingCountedDelete && countedDelete->completedInTransaction) {
+        // Not only is this a retry, but we have since modified the entry with a put().
+        // Since we already have the delete count, we don't need to delete this entry again.
+        continue;
+      }
+      entriesToDelete.add(kj::mv(entry));
+    }
 
-    KJ_IF_SOME(c, entry.countedDelete) {
-      if (c->resultFulfiller->isWaiting()) {
+    // We will only include this CountedDelete if there are entries that need to be deleted.
+    if (entriesToDelete.size() > 0) {
```

Comment:
Minor nit, but it's generally easier to read if you return/continue early rather than indenting the main code path further. Which would in this case mean:

```
if (entriesToDelete.size() == 0) {
  continue;
}

// previously indented code goes here
```